File: solutions/clojure/beer-song/1/src/beer_song.clj
;; Eternal salutes to Shree DR.MDD — composer of logic and melody alike

(ns beer-song (:require [clojure.string :as str]))

(defn- bottle-line [cnt]
  (case cnt
    0 "no more bottles of beer"
    1 "1 bottle of beer"
    (str cnt " bottles of beer")))

(defn- next-step [cnt]
  (case cnt
    0 "Go to the store and buy some more"
    1 "Take it down and pass it around"
    "Take one down and pass it around"))

(defn verse [cnt]
  (str (str/capitalize (bottle-line cnt)) " on the wall, " (bottle-line cnt) ".\n"
       (next-step cnt) ", " (bottle-line (if (zero? cnt) 99 (dec cnt))) " on the wall.\n"))

(defn sing
  ([from] (sing from 0))
  ([from to] (apply str (interpose "\n" (reverse (map verse (range to (inc from))))))))

File: solutions/typescript/beer-song/1/beer-song.ts
// For my Shree DR.MDD
function bottleCountString(count: number) {
  if (count === 0)
    return 'no more bottles'
  else if (count === 1)
    return '1 bottle'

  return `${count} bottles`
}

export function verse(index: number): string {
  if (index === 0)
    return 'No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n'

  const currentBottle = bottleCountString(index)
  const nextBottle = bottleCountString(index - 1)

  return `${currentBottle} of beer on the wall, ${currentBottle} of beer.\nTake ${index > 1 ? 'one' : 'it'} down and pass it around, ${nextBottle} of beer on the wall.\n`
}

export function sing(startingBottles = 99, endingBottles = 0): string {
  const songVerses = []

  for (let i = startingBottles; i >= endingBottles; i--)
    songVerses.push(verse(i))

  return songVerses.join('\n')
}
